generator/transit_generator.cpp

```diff
@@ -43,20 +59,82 @@ string GetFileName(string const & filePath)
+void DeserializeFromJson(my::Json const & root, string const & key, vector<Item> & items)
 {
-  vector<Item> items;
-
+  items.clear();
   DeserializerFromJson deserializer(root.get());
   deserializer(items, key.c_str());
+}
+
+void DeserializerGatesFromJson(my::Json const & root, string const & mwmDir, string const & countryId,
+                               vector<Gate> & gates)
+{
+  DeserializeFromJson(root, "gates", gates);
+
+  // Creating IndexRouter.
+  CountryFile const countryFile(countryId);
+  LocalCountryFile localCountryFile(mwmDir, countryFile, 0 /* version */);
+  localCountryFile.SyncWithDisk();
+  CHECK_EQUAL(localCountryFile.GetFiles(), MapOptions::MapWithCarRouting,
+              ("No correct mwm corresponding local country file:", localCountryFile,
+               ". Path:", my::JoinFoldersToPath(mwmDir, countryId + DATA_FILE_EXTENSION)));
+
+  Index index;
+  pair<MwmSet::MwmId, MwmSet::RegResult> const result = index.Register(localCountryFile);
+  CHECK_EQUAL(result.second, MwmSet::RegResult::Success, ());
+
+  Platform platform;
+  unique_ptr<storage::CountryInfoGetter> infoGetter =
+      storage::CountryInfoReader::CreateCountryInfoReader(platform);
+  CHECK(infoGetter, ());
+
+  auto const countryFileGetter = [&infoGetter](m2::PointD const & pt) {
+    return infoGetter->GetRegionCountryId(pt);
+  };
+
+  auto const getMwmRectByName = [&](string const & c) -> m2::RectD {
+    CHECK_EQUAL(countryId, c, ());
+    return infoGetter->GetLimitRectForLeaf(c);
+  };
 
-  if (handler)
-    handler(items);
+  auto const mwmId = index.GetMwmIdByCountryFile(countryFile);
+  CHECK_EQUAL(mwmId.GetInfo()->GetType(), MwmInfo::COUNTRY, ());
+  auto numMwmIds = make_shared<NumMwmIds>();
+  numMwmIds->RegisterFile(countryFile);
+
+  // Note. |indexRouter| is valid until |index| is valid.
+  IndexRouter indexRouter(VehicleType::Pedestrian, false /* load altitudes */,
+                          CountryParentNameGetterFn(), countryFileGetter, getMwmRectByName,
+                          numMwmIds, MakeNumMwmTree(*numMwmIds, *infoGetter),
+                          traffic::TrafficCache(), index);
+  unique_ptr<WorldGraph> worldGraph = indexRouter.MakeWorldGraph();
+  worldGraph->SetMode(WorldGraph::Mode::SingleMwm);
+
+  // Looking for the best segment for every gate.
+  bool dummy;
+  for (Gate & gate : gates)
+  {
+    // Note. For pedestrian routing all the segments are considered as twoway segments so
+    // IndexRouter.FindBestSegment() method finds the same segment for |isOutgoing| == true
+    // and |isOutgoing| == false.
+    Segment bestSegment;
+    if (indexRouter.FindBestSegment(gate.GetPoint(), m2::PointD::Zero() /* direction */,
+                                    true /* isOutgoing */, *worldGraph, bestSegment, dummy))
+    {
+      // Note. |numMwmIds| with only one mwm is used to create |indexRouter|. So |Segment::m_mwmId|
+      // is not valid at |bestSegment| and is not copied to |Gate::m_bestPedestrianSegment|.
```

Comment:
Why bestSegment.m_mwmId is not valid? It must be valid, no?

Reply:
Who did it write!?:) It should be a comment which explain why mwm id at |bestSegment| == 0.
Rewrote:
`// Note. |numMwmIds| with only one mwm is used to create |indexRouter|. Because of it |Segment::m_mwmId| == 0.`